### PR TITLE
Add 3D de-convolution

### PR DESCRIPTION
**Pull request by psyyz10**
## What changes were proposed in this pull request?
3D de-convolution for UCSF support.

## How was this patch tested?
Unit test

### CODE REVIEW

**Comment by SeaOfOcean** [comment]
miss createVolumetricConvolution in scala API?

**Comment by SeaOfOcean** [comment]
update markdown doc

**Comment by SeaOfOcean** [comment]
Add gradient check?

**Comment by psyyz10** [comment]
Sorry, forgot to commit last changes, have updated

File: pyspark/bigdl/nn/layer.py [hunk 1]
```diff
@@ -2969,6 +2969,99 @@ def set_init_method(self, weight_init_method = None, bias_init_method = None):
...
+    "Backwards Convolution," "Deconvolution", or "Upconvolution."
+
+
+    Reference Paper: Long J, Shelhamer E, Darrell T. Fully convolutional networks for semantic
+    segmentation[C]//Proceedings of the IEEE Conference on Computer Vision and Pattern Recognition.
+    2015: 3431-3440.
+
+    :param nInputPlane The number of expected input planes in the image given into forward()
+    :param nOutputPlane The number of output planes the convolution layer will produce.
+    :param kT The kernel width of the convolution.
```
Comment (SeaOfOcean) [inline pyspark/bigdl/nn/layer.py]: The kernel size of the convolution in time

File: pyspark/bigdl/nn/layer.py [hunk 2]
```diff
@@ -2969,6 +2969,99 @@ def set_init_method(self, weight_init_method = None, bias_init_method = None):
...
+    :param kW The kernel width of the convolution.
+    :param kH The kernel height of the convolution.
+    :param dT The step of the convolution in the width dimension. Default is 1.
```
Comment (SeaOfOcean) [inline pyspark/bigdl/nn/layer.py]: time

File: pyspark/bigdl/nn/layer.py [hunk 3]
```diff
@@ -2969,6 +2969,99 @@ def set_init_method(self, weight_init_method = None, bias_init_method = None):
...
+    :param dW The step of the convolution in the width dimension. Default is 1.
+    :param dH The step of the convolution in the height dimension. Default is 1.
+    :param padT The additional zeros added per width to the input planes. Default is 0.
```
Comment (SeaOfOcean) [inline pyspark/bigdl/nn/layer.py]: time

File: pyspark/bigdl/nn/layer.py [hunk 4]
```diff
@@ -2969,6 +2969,99 @@ def set_init_method(self, weight_init_method = None, bias_init_method = None):
...
+    :param padW The additional zeros added per width to the input planes. Default is 0.
+    :param padH The additional zeros added per height to the input planes. Default is 0.
+    :param adjT Extra depth to add to the output image. Default is 0.
+    :param adjW Extra width to add to the output image. Default is 0.
+    :param adjH Extra height to add to the output image. Default is 0.
+    :param nGroup Kernel group number.
+    :param noBias If bias is needed.
+    :param initMethod Init method, Default, Xavier, Bilinear.
```
Comment (SeaOfOcean) [inline pyspark/bigdl/nn/layer.py]: no initMethod

File: docs/docs/APIdocs/Layers/Convolution-Layers.md [hunk 5]
```diff
@@ -9,19 +9,27 @@ val m = SpatialConvolution(nInputPlane,nOutputPlane,kernelW,kernelH,strideW=1,st
 m = SpatialConvolution(n_input_plane,n_output_plane,kernel_w,kernel_h,stride_w=1,stride_h=1,pad_w=0,pad_h=0,n_group=1,propagate_back=True,wRegularizer=None,bRegularizer=None,init_weight=None,init_bias=None,init_grad_weight=None,init_grad_bias=None)
 ```
 
-SpatialConvolution is a module that applies a 2D convolution over an input image.
+SpatialConvolution is a module that applies a 3D full convolution over an 3D input image, a sequence of images, or a video etc. * The input tensor is expected to be a 4D or 5D(with batch) tensor. Note that instead
```
Comment (SeaOfOcean) [inline docs/docs/APIdocs/Layers/Convolution-Layers.md]: SpatialFullConvolution

Comment (SeaOfOcean) [inline docs/docs/APIdocs/Layers/Convolution-Layers.md]: seems you modified SpatialConvolution?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+import com.intel.analytics.bigdl.optim.Regularizer
+import com.intel.analytics.bigdl.tensor.{DenseTensorBLAS, DoubleType, FloatType, Tensor}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.reflect.ClassTag
+
+/**
+ * Apply a 3D full convolution over an input image.
+ *
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: duplicate

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+  var adjH: Int = 0,
+  val nGroup: Int = 1,
+  val noBias: Boolean = false,
+  var wRegularizer: Regularizer[T] = null,
+  var bRegularizer: Regularizer[T] = null
+)(implicit ev: TensorNumeric[T])
+  extends AbstractModule[Activity, Tensor[T], T] with Initializable {
+
+  require(adjW <= dW - 1 && adjH <= dH - 1 && adjT <= dT -1,
+    "VolumetricFullConvolution: adjW=$adjW and adjH=$adjH must be smaller than " +
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: miss "s" before string

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 8]
```diff
@@ -0,0 +1,848 @@
...
+      s"(dW - 1)=${dW - 1} and (dH - 1)=${dH - 1} respectively")
+
+  val weight: Tensor[T] = Tensor[T](nGroup, nInputPlane / nGroup,
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: val weight: Tensor[T] = if (initWeight != null) {
    initWeight
  } else {
    Tensor[T](xxxxxx)
  }

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 9]
```diff
@@ -0,0 +1,848 @@
...
+    nOutputPlane / nGroup, kT, kH, kW)
+  val bias: Tensor[T] = if (noBias) null else Tensor[T](nOutputPlane)
+
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: val bias: Tensor[T] = if (noBias) null
    else if (initBias != null) initBias else Tensor[T](nOutputPlane)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 10]
```diff
@@ -0,0 +1,848 @@
...
+  val gradWeight: Tensor[T] = Tensor[T](nGroup, nInputPlane / nGroup,
+    nOutputPlane / nGroup, kT, kH, kW)
+  val gradBias: Tensor[T] = if (noBias) null else Tensor[T](nOutputPlane)
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: val gradWeight: Tensor[T] = if (initGradWeight != null) {
    initGradWeight
  } else {
    Tensor[T](xxxx)
  }

  val gradBias: Tensor[T] = if (noBias) null
    else if (initGradBias != null) initGradBias else Tensor[T](nOutputPlane)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 11]
```diff
@@ -0,0 +1,848 @@
...
+  private val columns: Tensor[T] = Tensor[T]()
+  private val ones: Tensor[T] = Tensor[T]()
+  private val zeroScalar: Tensor[T] = Tensor[T]()
+  protected val onesBias = Tensor[T]()
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: if (noBias) null

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 12]
```diff
@@ -0,0 +1,848 @@
...
+  protected val onesBatch = Tensor[T]()
+  protected var weightMM: Tensor[T] = _
+  protected val gradientBiasMT: Tensor[T] = Tensor[T]()
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: if (noBias) null

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 13]
```diff
@@ -0,0 +1,848 @@
...
+    val outputDepth = (inputDepth - 1) * dT - 2 * padT + kT + adjT
+
+    // Batch size + input planes
+    val batchSize = inputTensor.size(1)
+
+    // Resize output
+    output.resize(batchSize, nOutputPlane, outputDepth, outputHeight, outputWidth)
+    output.zero()
+
+    if (onesBias.dim() != 1 || onesBias.size(1) != outputDepth * outputHeight * outputWidth) {
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: what if noBias

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 14]
```diff
@@ -0,0 +1,848 @@
...
+        "gradWeight" -> gradWeight, "gradBias" -> gradBias))
+    }
+  }
+
+  override def clearState() : this.type = {
+    super.clearState()
+    columns.set()
+    ones.set()
+    zeroScalar.set()
+    onesBias.set()
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: noBias

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 15]
```diff
@@ -0,0 +1,848 @@
...
+    hash = hash * seed + dH.hashCode()
+    hash = hash * seed + padW.hashCode()
+    hash = hash * seed + padH.hashCode()
+    hash = hash * seed + adjW.hashCode()
+    hash = hash * seed + adjH.hashCode()
+    hash = hash * seed + weight.hashCode()
+    hash = hash * seed + bias.hashCode()
+    hash = hash * seed + gradWeight.hashCode()
+    hash = hash * seed + gradBias.hashCode()
+
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: what if noBias

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 16]
```diff
@@ -0,0 +1,848 @@
...
+    hash
+  }
+
+  override def toString(): String = {
+    s"${getPrintName()}($nInputPlane -> $nOutputPlane, " +
+      s"$kW x $kH, $dW, $dH, $padW, $padH, $adjW, $adjH)"
+  }
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: toString need depth

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala (hunk 15 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: lack T

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 17]
```diff
@@ -0,0 +1,852 @@
...
+        s" Output size is too small")
+
+    require(input.nDimension() == ndim && input.size(dimFilter) == nInputPlane,
+      s"VolumetricFullConvolution: input's feature maps should be $nInputPlane, " +
+        s"but got ${input.size(dimFilter)}")
+
+    if (null != gradOutput) {
+      require(gradOutput.nDimension() == ndim, s"VolumetricFullConvolution: gradOutput should be " +
+        s"$ndim, but got ${gradOutput.nDimension()}")
+      require(gradOutput.size(dimFilter) == nOutputPlane
```
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: add check for gradOutput.size(dimDepth)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala [hunk 18]
```diff
@@ -0,0 +1,852 @@
...
+  }
+
+
+  override def accGradParameters(input: Activity, gradOutput: Tensor[T]): Unit = {
+    val inputTensor: Tensor[T] = if (input.isInstanceOf[Table]) {
+      val targetTensor: Tensor[T] = input.toTable[Tensor[T]](2)
+      val tDims = targetTensor.dim()
+      val tH = targetTensor.size(tDims - 1)
+      val tW = targetTensor.size(tDims)
+      adjW = calculateAdj(tW, kW, padW, dW)
```
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricFullConvolution.scala]: do we need to calculate adjT

**Pull request merged**